### PR TITLE
Feature/32 lambert module

### LINKED ISSUE #32: Lambert solver module
Create C++ FSW module that takes two position vectors, time of flight, number of orbits, and planet mu as inputs and outputs velocity vectors corresponding to the two input positions.

### PR DESCRIPTION
**Pull request by JulianHammerl**
* **Tickets addressed:** #32 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

Lambert solver module was developed using the Lambert algorithms by Gooding and Izzo.
The commits are organized the following way:
- Commits 1-3: create necessary message types for module
- Commit 4: Initialize module files
- Commit 5: Fill in Reset method
- Commit 6-17: Create necessary sub-functions for module
- Commit 18: Fill in UpdateState method
- Commit 19: Add UnitTest
- Commit 20: Add documentation
- Commit 21: Add release notes

## Verification

A UnitTest was added that compares the computed velocity vector solutions with true values obtained using first principles and an external Lambert solver python function.

## Documentation

This is a new module. An RST documentation was added.

## Future work

No future work anticipated at this time.

### CODE REVIEW

File: src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h
Comment (patkenneally) [inline src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h]: :large_blue_circle: Extra new line that can be removed.

File: src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h
Comment (patkenneally) [inline src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h]: :large_blue_circle: Underscores (unless representing frames etc.) don't conform with the Basilisk style guide. I think this should be v1Sol2. But I'm happy to hear rationale otherwise.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :small_red_triangle_down: Need a single new line at the end of the file. Notice the little red circle in the gutter on the left in the GitHub UI?

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: :small_red_triangle_down: Let's mark these as `override`, like `void Reset(uint64_t CurrentSimNanos) override;` (same for Update). It tells the compiler that we are explicitly overriding the abstract  and thee function signature should match exactly. If we don't and get the function signature slightly wrong it will compile and can lead to silent errors.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: I would declare and assign the variable in one line. What are you thoughts on that? Also you can remove the comments. They restate what the code shows.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: If the `mu` is bad, should we be assigning it to the member variable `this->mu`? Or should we check it first in the message, and only if good, assign it?

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :small_red_triangle_down: This needs to compare against an int `0`, rather than the float `0.0`.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: :small_red_triangle_down: Let's rename this to something like `numberOfRevolutions`. We can also change the bsklogger error message above to the same effect. As a user, being told "M is must be positive or zero", doesn't give me a clear knowledge about how to fix my issue. So changing the variable name would also fix that.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: :large_blue_circle: Extra new line can be removed.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: :large_orange_diamond: Are these going to hold multiple orbit frame definitions? Looking through the code it looks like a single orbit frame definition. if so, then we can simply use `Eigen::Vector3d` and forego the `std::vector`

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: Remove extra new line.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: This would be better as a std::array<double, 3>.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: Is there a neat place to add a line break into these two expressions such that they don't go over the desirable 120 character line length limit? 

I know I'm annoying :) and when we add in the automatic linting to the project these things will be checked by the build system and you'll be alerted. Until then, I'm here to bug you.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: Depending on what the solution is meant to be, you could also directly return a std::array<double, 3> here.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: Declare the incremented variable inside the for loop statement.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: I would normally say let's declare this as const above, but we know we are going to pull these methods out into some kind of support lib so let's keep it all together for now.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: Maybe std::array<double, 3> again?

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: std::array here also?

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: std::array?

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :small_red_triangle_down: Let's split out this logic into two methods. Something like a `goodingInitialGuess` and an `izzoInitialGuess`. This pulls the method logic up out of the method which claims to get the guess and to the place where we should be deciding on which guess to get. It also greatly simplifies the cognitive load of this method making it easier to understand and maintain in the future. Finally, it also increases test-ability. While we don't have it now, I am integrating Google Test for direct C++ testing in Basilisk. These methods are exactly methods we need to test, yet don't want to expose to the Python user API. So testing at the C++ level is what we want. Separating these components out will allow us to write tests for each method individually. The test code will also be simpler because of it.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_orange_diamond: std::array ?

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp:198]: :small_red_triangle_down: Same as before. Let's split these out to a gooding method and an izzo method. This now allows us to localize the logic for switching between methods to one place in the module. Makes it easier to follow, maintain, and test.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (patkenneally) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: :large_blue_circle: Extra new line can be deleted.

**Review by patkenneally — CHANGES_REQUESTED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

@JulianHammerl this is a really good PR. Nicely done. Solid job. Happy to chat about any of my comments. I've also yet to study the tests but wanted to share this feedback now so you can get started on it if you like. I'm going to look at the test over the next 24 hrs.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (thibaudteil) [inline src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: 🔶 The metric for alignment seems too tight, if these vectors are coming from estimated states they might be noisy in which case we would want a broader exclusion to prevent nearly collinear vectors from giving a solution. We could make this larger by default, perhaps equivalent to a degree, and add it as a parameter?

File: src/architecture/msgPayloadDefC/LambertPerformanceMsgPayload.h
Comment (thibaudteil) [inline src/architecture/msgPayloadDefC/LambertPerformanceMsgPayload.h]: 🔵 Just pointing out a couple more vectors with underscores echoing Pat's Basilisk style guide comment. Sorry for not catching this sooner.

File: src/fswAlgorithms/orbitControl/lambertSolver/_UnitTest/test_lambertSolver.py
Comment (thibaudteil) [inline src/fswAlgorithms/orbitControl/lambertSolver/_UnitTest/test_lambertSolver.py:75]: 🔵 Just for extra clarity, not at all necessary to modify: you could emphasize the fact that you are testing the solver in two ways. The first is using external orbit dynamics libraries and results, the second is using a python implementation of the algorithm. This ensures that results are physical independently of the algorithm, but also that the Cpp algorithm matches the python implementation. It's a good test!

**Review by thibaudteil — APPROVED** [review]
Nice work Julian, great addition

File: src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h
Comment (JulianHammerl) [reply src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h]: Done.

File: src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h
Comment (JulianHammerl) [reply src/architecture/msgPayloadDefC/LambertSolutionMsgPayload.h]: Done.

File: src/architecture/msgPayloadDefC/LambertPerformanceMsgPayload.h
Comment (JulianHammerl) [reply src/architecture/msgPayloadDefC/LambertPerformanceMsgPayload.h]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Sounds good.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done. Added variable numIter to count number of iterations instead.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp:198]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/_UnitTest/test_lambertSolver.py
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/_UnitTest/test_lambertSolver.py:75]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done. Also changed it for the remaining message variables that are checked for validity.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.h]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done.

File: src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp
Comment (JulianHammerl) [reply src/fswAlgorithms/orbitControl/lambertSolver/lambertSolver.cpp]: Done. I added an optional module input parameter called alignmentThreshold that defaults to 1.0 degrees if not specified. I also updated the unit test to test this new parameter.

**Review by patkenneally — APPROVED** [review]

**Pull request merged**
